Doctrine to WPDB - timezone fixes [MAILPOET-6142]

costasovo opened the pull request:
## Description

This PR fixes issues related to timezones. With the WPDB refactor, we can't no longer rely on the timezone set on the DB connection. This change was causing a number of issues.

1. We started presenting times in listings in GMT
2. Scheduling was using incorrect times

During fixing I found that we use a couple of different approaches so I unified those as outlined [in the related P2 post](https://mailpoetdev.wordpress.com/2024/08/07/rfc-timezones-in-the-plugin-unification/).

The rules are simple.

- In PHP and DB we store times in GMT
- When presenting times to users or when we want them to enter time we expect site's timezone

## Code review notes

Please review commit by commit. The changes and motivations are described in the commit message or (I hope) are obvious.

## QA notes

This will be merged with the WPDB branch. I'll update testing instructions there. 

## Linked PRs

_N/A_

## Linked tickets

 [MAILPOET-6142]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] ⛔ I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes - I was doing some batch refactors with IDE so I touched some jsx files but didn't rewrite them to Typescript.


[MAILPOET-6142]: https://mailpoet.atlassian.net/browse/MAILPOET-6142?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Use UTC time in TimestampListener

$this->wp->currentTime('timestamp') adds 'gmt_offset' value.
This was needed with PDO connection because the connection
used @session.time_zone set to the same offset value.

[MAILPOET-6142]

costasovo committed:
Improve LastSubsribedAtListener to use fresh now

The cached now may cause the value to not be exact.
[MAILPOET-6142]

costasovo committed:
Replace usage of WP's current_time without gmt parameter in libs

This commit replaces usages by Carbon::now or in case we need a timestamp it
keeps current_time but adds the gtm parameter as true.
[MAILPOET-6142]

costasovo committed:
Replace current_time usage in tests

[MAILPOET-6142]

costasovo committed:
Pass GMT offset from WordPress settings to JS

[MAILPOET-6142]

costasovo committed:
Add formatFromGmt etc to MailPoet.Date

This set of function expects input in GMT and add site's offset
They are meant to be used for data coming from server in GMT
[MAILPOET-6142]

costasovo committed:
Replace simple usages of date formating with offet aware methods

[MAILPOET-6142]

costasovo committed:
Use proper time in scheduling calendar on send page

We add offset before we display time and we remove when we pass it to callback.
[MAILPOET-6142]

costasovo committed:
Show proper times for action scheduler info on help page

[MAILPOET-6142]

costasovo committed:
Use isInFutureGmt to check future on send page

[MAILPOET-6142]

costasovo committed:
Show proper time in listing for scheduled newsletter

[MAILPOET-6142]

costasovo committed:
Fix times in task lists on Help > System Status

The time adjustments are not necessary because we rely on WP setting
timezone to UTC.
Manipulation in adjustTimezone difference was causing us to display GMT time.
[MAILPOET-6142]

costasovo committed:
Fix values in list of times in scheduling form

[MAILPOET-6142]

costasovo committed:
Fix scheduling post notifications

The user picks time in a select box, and we expect they pick a time
in the site's timezone so the saved cron expression is in the site's timezone.
When we calculate the next run date, we do it in the site's timezone and
to get UTC time, we convert it.
[MAILPOET-6142]

costasovo committed:
Set GMT timezone to WPDB

[MAILPOET-6142]

costasovo committed:
Remove adjustForTimezoneDifference functionality

This is no longer used. It was used in just a couple of places
and was fixing a difference between PHP time zone and browser time.
We rely on WordPress to set PHP timezone to UTC and we use the site's
timezone on the client so the fix is no longer useful.
[MAILPOET-6142]

JanJakes commented on mailpoet/lib/Newsletter/Sending/ScheduledTasksRepository.php:
```diff
@@ -311,7 +307,7 @@ public function findByTypeAndSubscriber(string $type, SubscriberEntity $subscrib
 
   public function touchAllByIds(array $ids): void {
-    $now = CarbonImmutable::createFromTimestamp((int)$this->wp->currentTime('timestamp'));
+    $now = CarbonImmutable::now();
```
This one doesn't have the `millisecond(0)` part. Should we add it, for consistency?

JanJakes commented on mailpoet/assets/js/src/date.ts:
```diff
@@ -14,12 +14,19 @@ export const MailPoetDate: {
   isInFuture: (dateString: string, currentTime: MomentInput) => boolean;
+  isInFutureGmt: (dateString: string, currentTime: MomentInput) => boolean;
+  isInPast: (dateString: string, currentTime: MomentInput) => boolean;
+  isInPastGmt: (dateString: string, currentTime: MomentInput) => boolean;
```
Let's keep only one variant here (the new one) to avoid confusion when using the Date object.

JanJakes commented on mailpoet/assets/js/src/date.ts:
```diff
@@ -120,6 +121,19 @@ export const MailPoetDate: {
+      format: 'Y-m-d H:i:s',
+      offset: parseFloat(window.mailpoet_date_offset || '0') * -1,
+    });
+  },
```
As both the input and output seem to be strings and I guess they both are `Y-m-d H:i:s`, I'm not sure if the `timestamp` in the name makes sense. Maybe we could use a `string` suffix, or sth like `datetime`.

Actually, I think the first method should follow the naming of the other methods like `fullFromGmt` (and after simplifying to only one variant, it would be only `full`), so I guess that format needs a name. We could name it `datetimeString`, maybe, to avoid confusion with datetime objects.

As for the second method, I think we first need a generic counterpart to `formatFromGmt` (~ `format`). Something like `formatToGmt`, maybe. Then, if the first method is named `datetimeString`, then the second one could be called `toGmtDatetimeString`, or something like that. And I suppose both could accept any date input, not just strings, to align with the rest of the API.

JanJakes commented on mailpoet/assets/js/src/newsletters/send/standard.tsx:
```diff
@@ -85,7 +83,8 @@ class StandardScheduling extends Component<StandardSchedulingProps> {
           <span className="mailpoet-form-schedule-time">
-            {__('Your website’s time is', 'mailpoet')} {currentTime}
+            {__('Your website’s time is', 'mailpoet')}{' '}
+            {MailPoet.Date.timeFromGmt(new Date().toISOString())}
```
We've tested that this should work correctly. Is the `toISOString()` part needed?

JanJakes replied:
Oh, and seeing `Your website’s time is`, I wonder if it would make sense to add such information to somewhere to Help and maybe also the timezone offset. That might be out of the scope of this PR, however.

costasovo replied:
Website time is visible in WP's settings, but we didn't have WP's timezone in the system report (also sent to the support), so I added it there.

JanJakes commented on mailpoet/assets/js/src/newsletters/send/standard.tsx:
```diff
@@ -239,13 +238,12 @@ export const StandardNewsletterFields = {
+      MailPoet.Date.isInFutureGmt(
         newsletter.options?.scheduledAt,
-        currentDateTime,
+        new Date().toISOString(),
```
Is the `toISOString()` needed here?

JanJakes commented on mailpoet/assets/js/src/common/listings/_stories/newsletter-status.tsx:
```diff
@@ -24,20 +24,24 @@ export function NewsletterStatuses() {
+      <NewsletterStatus scheduledFor={inSixHours.toISOString()} />
+      <NewsletterStatus scheduledFor={inTwelveHours.toISOString()} />
+      <NewsletterStatus scheduledFor={inOneDay.toISOString()} />
+      <NewsletterStatus scheduledFor={inOneDay.toISOString()} isPaused />
```
Actually, I'm realizing that my question applies more generally — is `.toISOString()` needed anywhere at all? If the new datetime methods accept also date objects and take their own timezone into consideration rather than stripping it, then this should not be necessary.

costasovo replied:
In this case, this is just for the story book. The newsletter status component newly accepts only string value (as it used to in the past) and handles dates formating internally. The .toISOString was the simplest way how to fix typescript warnings comming from the stories.

JanJakes commented on mailpoet/lib/Newsletter/Scheduler/Scheduler.php:
```diff
@@ -33,11 +33,16 @@ public function getNextRunDate($schedule, $fromTimestamp = false) {
   public function getPreviousRunDate($schedule, $fromTimestamp = false) {
-    $fromTimestamp = ($fromTimestamp) ?: $this->wp->currentTime('timestamp', true);
+    // User enters time in WordPress site timezone, but we need to calculate it in UTC before we save it to DB
+    // 1) As the initial tile we use time in site timezone (gmt is false)
```
`tile` is probably a typo

JanJakes commented on mailpoet/lib/Newsletter/Scheduler/Scheduler.php:
```diff
@@ -33,11 +33,16 @@ public function getNextRunDate($schedule, $fromTimestamp = false) {
+    // 1) As the initial tile we use time in site timezone (gmt is false)
+    // 2) We use CronExpression to calculate previous run (still in site's timezone)
+    // 3) We convert the calculated time to UTC
+    $fromTimestamp = ($fromTimestamp) ?: $this->wp->currentTime('timestamp', false);
```
I suppose this whole expression could be replaced by a more understandable `$from = $this->wp->currentDatetime()` that would call WP's `current_datetime` function.

JanJakes commented on mailpoet/lib/Newsletter/Scheduler/Scheduler.php:
```diff
@@ -82,10 +87,15 @@ public function formatDatetimeString($datetimeString) {
+    // 1) As the initial tile we use time in site timezone (gmt is false)
+    // 2) We use CronExpression to calculate next run (still in site's timezone)
+    // 3) We convert the calculated time to UTC
+    $fromTimestamp = $fromTimestamp ?: $this->wp->currentTime('timestamp', false);
```
Same applies as on lines 37 & 40.

JanJakes commented on mailpoet/lib/Doctrine/WPDB/Connection.php:
```diff
@@ -18,6 +18,11 @@ public function __construct() {
+     * Ensure connection is in UTC timezone.
+     * We use timestamp columns and we expect to get UTC time
+     */
+    $wpdb->query("SET time_zone = '+00:00'");
```
I think that currently the WPDB driver doesn't include much MailPoet-specific functionality. This line seems more specific, so I'm just wondering if it could be handled [via a DBAL event](https://www.doctrine-project.org/projects/doctrine-dbal/en/3.1/reference/events.html#postconnect-event), just to keep it separated. It would also go through the proper `query()` error handling. Probably not a big deal, just an idea.

JanJakes reviewed (commented):
This is a great piece of work, thanks for handling this! 👏

costasovo committed:
Use DBAL Driver middleware to handle the correct connection timezone

I looked into using DBAL events for this but found that DBAL events
are deprecated so I went with middleware as they are recommended
as a replacement for the DBAL events.
[MAILPOET-6142]

costasovo committed:
Remove unused date.ts methods and rename the prefixed methods

There is no use case for the methods without the translation to site's timezone.
When the input doesn't specify the timezone we treat it as utc.
When the input provides timezone (e.g. '2024-08-13 10:00+02:00' or new Date('2024-08-13 10:00+02:00'))
We respect the timezone of the input and print the result in the site's timezone.

When the site's timezone is set +01:00

'2024-08-13 10:00' prints as 2024-08-13 11:00
'2024-08-13 10:00:+02:00' prints as 2024-08-13 09:00
new Date('2024-08-13 10:00:+02:00') prints as 2024-08-13 09:00
new Date('2024-08-13 10:00:+00:00') prints as 2024-08-13 11:00

[MAILPOET-6142]

costasovo committed:
Replace usages of prefixed date methods with renamed variants

[MAILPOET-6142]

costasovo committed:
Get rid of unnecessary usages of .toISOString

[MAILPOET-6142]

costasovo committed:
Rename methods for manipulating date in scheduling UI

[MAILPOET-6142]

costasovo committed:
Get rid of unnecessary parameter in the Scheduler

The fromTimestamp parameter was used only in tests.
This commit replaces the parameter in tests by mocks and removes it.
 [MAILPOET-6142]

costasovo committed:
Improve readability of scheduler code

[MAILPOET-6142]

costasovo committed:
Add timezone info to system report

 [MAILPOET-6142]

costasovo commented:
@JanJakes Thank you for the detailed review and the suggestions; they improved the final outcome. I believe I addressed all of them. I addressed them in 8 new commits. Could you please take another look?

JanJakes commented on mailpoet/lib/Doctrine/Middlewares/PostConnectMiddleware.php:
```diff
@@ -0,0 +1,14 @@
+class PostConnectMiddleware extends AbstractDriverMiddleware {
+  public function connect(array $params): Connection {
+    $connection = parent::connect($params);
+    $res = $connection->exec('SET time_zone = "+00:00"');
```
`$res` seems to be unused (but no CI checks are complaining).

costasovo replied:
You're right; I used it to debug the return value. But strange there are no warnings

JanJakes commented on mailpoet/assets/js/src/date.ts:
```diff
@@ -21,8 +21,8 @@ export const MailPoetDate: {
-  gmtToSiteTimestamp: (utcTimeStamp: string) => string;
-  siteToGmtTimestamp: (siteTimeStamp: string) => string;
+  datetimeString: (utcTimeStamp: string) => string;
+  toGmtDatetimeString: (siteTimeStamp: string) => string;
```
Probably forgotten, I suppose here the parameters should be `date: MomentInput`.

JanJakes reviewed (commented):
Great, thanks! Just two tiny comments.
